src/script/interpreter.cpp

```diff
@@ -2041,6 +2042,148 @@ bool GenericTransactionSignatureChecker<T>::CheckSequence(
+                script_pubkey.begin() + TAPROOT_SIZE_WITHOUT_STATE);
+
+    if (stack.size() == 0) {
+        return set_error(serror, ScriptError::INVALID_STACK_OPERATION);
+    }
+    if (stack.size() >= 2 && !stack.back().empty() &&
+        stack.back()[0] == TAPROOT_ANNEX_TAG) {
+        return set_error(serror, ScriptError::TAPROOT_ANNEX_NOT_SUPPORTED);
+    }
+    if (stack.size() == 1) {
+        // Spend using single signature instead of executing script
+        valtype &vch_sig = stacktop(-1);
+        bool f_success = false;
+        if (!EvalChecksig(vch_sig, vch_pubkey, script_pubkey.begin(),
+                          script_pubkey.end(),
+                          flags | SCRIPT_TAPROOT_KEY_SPEND_PATH, checker,
+                          metrics, std::nullopt, serror, f_success)) {
+            if (serror && *serror == ScriptError::SIG_NULLFAIL) {
+                *serror = ScriptError::TAPROOT_VERIFY_SIGNATURE_FAILED;
+            }
+            // serror is set
+            return false;
+        }
+        if (!f_success) {
+            return set_error(serror,
+                             ScriptError::TAPROOT_VERIFY_SIGNATURE_FAILED);
+        }
+        return set_success(serror);
+    }
+    // Spend using executing script, internal pubkey and merkle path
+    valtype control_block = stacktop(-1);
+    valtype script_bytes = stacktop(-2);
+    CScript exec_script(script_bytes.begin(), script_bytes.end());
+    popstack(stack);
+    popstack(stack);
+    const uint32_t size_remainder =
+        (control_block.size() - TAPROOT_CONTROL_BASE_SIZE) %
+        TAPROOT_CONTROL_NODE_SIZE;
+    if (control_block.size() < TAPROOT_CONTROL_BASE_SIZE ||
+        control_block.size() > TAPROOT_CONTROL_MAX_SIZE ||
+        (size_remainder != 0)) {
+        return set_error(serror, ScriptError::TAPROOT_WRONG_CONTROL_SIZE);
+    }
+    if ((control_block[0] & TAPROOT_LEAF_MASK) != TAPROOT_LEAF_TAPSCRIPT) {
+        return set_error(serror,
+                         ScriptError::TAPROOT_LEAF_VERSION_NOT_SUPPORTED);
+    }
+    uint256 tapleaf_hash;
+    if (!VerifyTaprootCommitment(tapleaf_hash, control_block, vch_pubkey,
+                                 exec_script)) {
+        return set_error(serror, ScriptError::TAPROOT_VERIFY_COMMITMENT_FAILED);
+    }
+    if (script_pubkey.size() == TAPROOT_SIZE_WITH_STATE) {
+        stack.push_back(
+            valtype(script_pubkey.begin() + TAPROOT_SIZE_WITHOUT_STATE + 1,
+                    script_pubkey.begin() + TAPROOT_SIZE_WITH_STATE));
+    }
+    ScriptExecutionData execdata{tapleaf_hash};
+    if (!EvalScript(stack, exec_script, flags, checker, metrics, execdata,
+                    serror)) {
+        // serror is set
+        return false;
+    }
+    if (stack.empty() || CastToBool(stack.back()) == false) {
+        return set_error(serror, ScriptError::EVAL_FALSE);
+    }
+    if (!VerifyScriptPostConditions(stack, script_sig, flags, metrics,
```

Comment:
I get the sigcheck asserts, however it would be better to maintain a counter in the EvalScript metrics and fail prematurely rather than continuing to execute them. I suppose this is okay as we won't be able to get spammed with failing transactions though.